manager/src/main/java/org/openremote/manager/dashboard/DashboardResourceImpl.java

```diff
@@ -28,59 +54,47 @@ public DashboardResourceImpl(TimerService timerService, ManagerIdentityService i
-            if(isAuthenticated()) {
-                publicOnly = (!hasResourceRole(ClientRole.READ_INSIGHTS.getValue(), Constants.KEYCLOAK_CLIENT_ID));
-            }
-            return this.dashboardStorageService.query(null, realm, getUserId(), publicOnly, false);
-
-        } catch (IllegalStateException ex) {
-            ex.printStackTrace();
-            throw new WebApplicationException(ex, INTERNAL_SERVER_ERROR);
+        if(!isRealmActiveAndAccessible(realm)) {
+            throw new WebApplicationException(FORBIDDEN);
         }
+        return dashboardStorageService.query(this.createDashboardQuery(realm), getUserId());
     }
 
     @Override
     public Dashboard get(RequestParams requestParams, String realm, String dashboardId) {
-        boolean publicOnly = true;
         if(realm == null) {
             throw new WebApplicationException(BAD_REQUEST);
         }
-        try {
-            // Realm should be enabled. Also takes unauthenticated users into count.
-            if(!isRealmActiveAndAccessible(realm)) {
+        Dashboard[] dashboards = dashboardStorageService.query(
+                this.createDashboardQuery(realm).ids(dashboardId).limit(1),
+                getUserId()
+        );
+        // If no dashboards were returned, check whether it existed, and return a different response.
+        if(dashboards.length == 0) {
+            if(this.dashboardStorageService.exists(dashboardId, realm)) {
                 throw new WebApplicationException(FORBIDDEN);
+            } else {
+                throw new WebApplicationException(NOT_FOUND);
             }
-            // if not authenticated, or having no INSIGHTS access, only take public dashboards into count
-            if(isAuthenticated()) {
-                publicOnly = (!hasResourceRole(ClientRole.READ_INSIGHTS.getValue(), Constants.KEYCLOAK_CLIENT_ID));
-            }
-            Dashboard[] dashboards = this.dashboardStorageService.query(Collections.singletonList(dashboardId), realm, getUserId(), publicOnly, false);
-            if(dashboards.length == 0) {
-                if(this.dashboardStorageService.exists(dashboardId, realm)) {
-                    throw new WebApplicationException(FORBIDDEN); // when no dashboard returned from query, but it does exist.
-                } else {
-                    throw new WebApplicationException(NOT_FOUND); // aka it does not exist
-                }
-            }
-            return dashboards[0]; // only return first entry
+        }
 
-        } catch (IllegalStateException ex) {
-            ex.printStackTrace();
-            throw new WebApplicationException(ex, INTERNAL_SERVER_ERROR);
+        return dashboards[0];
+    }
+
+    @Override
+    public Dashboard[] query(RequestParams requestParams, DashboardQuery query) {
+        if(query == null) {
+            query = this.createDashboardQuery(getRequestRealmName());
```

Comment:
`getRequestRealmName()` only looks at the request but this may not be the user's realm and should only be allowed for super users; normal users should just look at authenticated realm.

We should probably change the `getRequestRealmName()` so this mistake cannot be made as it is too easy to make currently.

Reply:
I was aware of differences between  `getRequestRealmName()` and `getAuthenticatedRealmName()` ;)
This is using the request realm on purpose, as the `/query` endpoint is also used for querying public dashboard that don't require authorization.

Will double check whether the cross-realm access is covered in tests right now.